src/python_testing/TC_VALCC_3_1.py

```diff
@@ -75,24 +78,26 @@ async def test_TC_VALCC_3_1(self):
             pass
 
         self.step(3)
-        target_state_dut = await self.read_valcc_attribute_expect_success(endpoint=endpoint, attribute=attributes.TargetState)
-
-        asserts.assert_true(target_state_dut is not NullValue, "TargetState is null")
-        asserts.assert_equal(target_state_dut, Clusters.Objects.ValveConfigurationAndControl.Enums.ValveStateEnum.kOpen,
-                             "TargetState is not the expected value")
 
-        self.step(4)
+        # Read target state first in case the current state goes to open between these calls.
+        # The test re-reads target state after CurrentState goes to open.
+        target_state_dut = await self.read_valcc_attribute_expect_success(endpoint=endpoint, attribute=attributes.TargetState)
         current_state_dut = await self.read_valcc_attribute_expect_success(endpoint=endpoint, attribute=attributes.CurrentState)
-        asserts.assert_true(current_state_dut is not NullValue, "CurrentState is null")
 
-        while current_state_dut is Clusters.Objects.ValveConfigurationAndControl.Enums.ValveStateEnum.kTransitioning:
+        while current_state_dut is cluster.Enums.ValveStateEnum.kTransitioning:
+            asserts.assert_equal(target_state_dut, cluster.Enums.ValveStateEnum.kOpen, "TargetState is incorrect")
             time.sleep(1)
 
+            target_state_dut = await self.read_valcc_attribute_expect_success(endpoint=endpoint, attribute=attributes.TargetState)
             current_state_dut = await self.read_valcc_attribute_expect_success(endpoint=endpoint, attribute=attributes.CurrentState)
             asserts.assert_true(current_state_dut is not NullValue, "CurrentState is null")
+            asserts.assert_true(target_state_dut is not NullValue, "TargetState is null")
```

Comment:
should these nullability asserts also apply outside of the while? It seems like conditions should always be consistent except we have a `retry while current state is transitioning`.

Reply:
I believe the checks are also done in step 4, it just means we read it twice to get there. Will add a checkmark, up to you if this really needs update or not.

Reply:
I ended up having to refactor this thing a bunch more, and this code is now gone. But the code is quite different from when you last looked. Would you mind taking another peek when you get a chance?